File: kotlin-java/src/main/kotlin/programmers/lv2/미로_탈출.java
package programmers.lv2;

import java.util.*;

public class 미로_탈출 {
    public static void main(String[] args) {

    }

    class Node {
        int x;
        int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    int[] dx = {0, 0, 1, -1};
    int[] dy = {1, -1, 0, 0};
    int n, m;
    int startX = -1;
    int startY = -1;

    public int solution(String[] maps) {

        n = maps.length;
        m = maps[0].length();

        for (int i = 0; i < n; i++) {
            boolean finished = false;

            for (int j = 0; j < m; j++) {
                if (maps[i].charAt(j) == 'S') {
                    finished = true;
                    startX = i;
                    startY = j;
                    break;
                }
            }
            if (finished) break;
        }

        int leverDistance = bfs(maps, 'L');
        if (leverDistance == -1) return -1;

        int exitDistance = bfs(maps, 'E');
        if (exitDistance == -1) return -1;

        return leverDistance + exitDistance;
    }


    int bfs(String[] maps, char goal) {
        int[][] visited = new int[n][m];

        visited[startX][startY] = 1;

        Queue<Node> q = new LinkedList<Node>();
        q.offer(new Node(startX, startY));

        while (!q.isEmpty()) {
            Node target = q.poll();

            for (int i = 0; i < 4; i++) {
                int nx = target.x + dx[i];
                int ny = target.y + dy[i];

                if (nx < 0 ||
                        nx >= n ||
                        ny < 0 ||
                        ny >= m ||
                        visited[nx][ny] != 0 ||
                        maps[nx].charAt(ny) == 'X') continue;

                visited[nx][ny] = visited[target.x][target.y] + 1;
                q.offer(new Node(nx, ny));

                if (maps[nx].charAt(ny) == goal) {
                    startX = nx;
                    startY = ny;
                    return visited[nx][ny] - 1;
                }
            }
        }
        return -1;
    }
}

File: kotlin-java/src/main/kotlin/baekjoon/gold/g4/미로 탈출/미로 탈출.kt
package baekjoon.gold.g4.`미로 탈출`

import java.util.LinkedList
import java.util.Queue
import java.util.StringTokenizer

data class Node(val x: Int, val y: Int, val wallState: Int)

val dx = intArrayOf(0, 0, 1, -1)
val dy = intArrayOf(1, -1, 0, 0)

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val (startX, startY) = readLine().split(" ").map { it.toInt() - 1 }
    val (endX, endY) = readLine().split(" ").map { it.toInt() - 1 }
    val map = Array(n) {
        val st = StringTokenizer(readLine())
        List(m) { st.nextToken().toInt() }
    }

    val visited = Array(n) { Array(m) { Array(2) { 0 } } }
    val q = LinkedList<Node>() as Queue<Node>
    visited[startX][startY][0] = 1
    q.offer(Node(startX, startY, 0))

    while (q.isNotEmpty()) {
        val cur = q.poll()

        for (i in 0..<4) {
            val nx = cur.x + dx[i]
            val ny = cur.y + dy[i]

            if (nx !in 0..<n || ny !in 0..<m) continue

            if (nx == endX && ny == endY) {
                println(visited[cur.x][cur.y][cur.wallState])
                return
            }

            if (map[nx][ny] == 1) {
                if (cur.wallState == 0 && visited[nx][ny][1] == 0) {
                    visited[nx][ny][1] = visited[cur.x][cur.y][0] + 1
                    q.offer(Node(nx, ny, 1))
                }
            } else {
                if (visited[nx][ny][cur.wallState] == 0) {
                    visited[nx][ny][cur.wallState] = visited[cur.x][cur.y][cur.wallState] + 1
                    q.offer(Node(nx, ny, cur.wallState))
                }
            }
        }
    }

    println(-1)
}